// @cool-vue/unix/src/tailwind/css.ts
import valueParser from "postcss-value-parser";
import type { Plugin } from "vite";
import { remToPx, rgbToRgba, toSafeClass } from "./utils";

export function cssPlugin(): Plugin {
	return {
		name: "vite-cool-unix-tailwind-css",
		enforce: "pre",

		config() {
			return {
				css: {
					postcss: {
						plugins: [
							{
								postcssPlugin: "vite-cool-uniappx-class-mapping",
								prepare() {
									return {
										// 处理选择器规则
										Rule(rule: any) {
											if (
												[
													".button-hover",
													":deep(",
													"&::",
													"uni-",
													".uni-"
												].some((e) => rule.selector.includes(e))
											) {
												return;
											}

											// 转换选择器为安全的类名格式
											rule.selector = toSafeClass(rule.selector);
										},

										// 处理声明规则
										Declaration(decl: any) {
											const className = decl.parent.selector || "";

											if (!decl.parent._twValues) {
												decl.parent._twValues = {};
											}

											// 处理 Tailwind 自定义属性
											if (decl.prop.includes("--tw-")) {
												decl.parent._twValues[decl.prop] =
													decl.value.includes("rem")
														? remToPx(decl.value)
														: decl.value;

												decl.remove();
												return;
											}

											// 转换 RGB 颜色为 RGBA 格式
											if (
												decl.value.includes("rgb(") &&
												decl.value.includes("/")
											) {
												decl.value = rgbToRgba(decl.value);
											}

											// 处理文本大小相关样式
											if (
												decl.value.includes("px") &&
												decl.prop == "color" &&
												className.includes("text-")
											) {
												decl.prop = "font-size";
											}

											// 删除不支持的属性
											if (["filter"].includes(decl.prop)) {
												decl.remove();
												return;
											}

											// 处理 flex-1
											if (decl.prop == "flex") {
												if (decl.value.startsWith("1")) {
													decl.value = "1";
												}
											}

											// 处理 vertical-align 属性
											if (decl.prop == "vertical-align") {
												decl.remove();
											}

											// 处理 visibility 属性
											if (decl.prop == "visibility") {
												decl.remove();
											}

											// 处理 sticky 属性
											if (className == ".sticky") {
												if (
													decl.prop == "position" ||
													decl.value == "sticky"
												) {
													decl.remove();
												}
											}

											// 解析声明值
											const parsed = valueParser(decl.value);
											let hasChanges = false;

											// 遍历并处理声明值中的节点
											parsed.walk((node: any) => {
												// 处理单位转换(rem -> px)
												if (node.type === "word") {
													const unit = valueParser.unit(node.value);

													if (typeof unit != "boolean") {
														if (unit?.unit === "rem") {
															node.value = remToPx(unit.number);
															hasChanges = true;
														}
													}
												}

												// 处理 CSS 变量
												if (
													node.type === "function" &&
													node.value === "var"
												) {
													const twKey = node.nodes[0]?.value;

													// 替换 Tailwind 变量为实际值
													if (twKey?.startsWith("--tw-")) {
														if (decl.parent._twValues) {
															node.type = "word";
															node.value =
																decl.parent._twValues[twKey] ||
																"none";

															hasChanges = true;
														}
													}
												}
											});

											// 更新声明值
											if (hasChanges) {
												decl.value = parsed.toString();
											}

											// 移除 Tailwind 生成的无效 none 变换
											const nones = [
												"translate(none, none)",
												"rotate(none)",
												"skewX(none)",
												"skewY(none)",
												"scaleX(none)",
												"scaleY(none)"
											];

											if (decl.value) {
												nones.forEach((noneStr) => {
													decl.value = decl.value.replace(noneStr, "");

													if (!decl.value || !decl.value.trim()) {
														decl.value = "none";
													}
												});
											}
										}
									};
								}
							}
						]
					}
				}
			};
		}
	};
}
