// @cool-vue/unix@1.1.0-beta.6/src/code/index.ts
import type { Plugin } from "vite";
import { getCodeIndex, rootDir } from "../utils";
import { readdirSync } from "fs";
import { basename } from "path";
import { globSync } from "glob";

function jsonPlugin(): Plugin {
	return {
		name: "vite-cool-unix-json",
		transform(code, id) {
			if (id.endsWith(".json")) {
				return {
					code: code.replace("new UTSJSONObject", ""),
					map: { mappings: "" }
				};
			}
		}
	};
}

function tsPlugin(): Plugin {
	return {
		name: "vite-cool-unix-ts",
		enforce: "pre",
		transform(code, id) {
			// 处理 main.ts 或 main.uts 文件，自动注入插件导入和使用代码
			if (id.endsWith("/main.ts") || id.endsWith("/main.uts")) {
				const pluginDirectory = rootDir("plugins");

				let pluginFiles: any[] = [];

				try {
					// 读取插件目录下的所有文件
					pluginFiles = readdirSync(pluginDirectory);
				} catch (error) {
					// 忽略错误
				}

				// 提取插件文件名（去除 .ts 扩展名）
				const pluginNames = pluginFiles.map((file) => basename(file as string, ".ts"));

				// 生成插件导入语句
				const importStatements = pluginNames.map((pluginName) => {
					return `import __${pluginName}Plugin from "@/plugins/${pluginName}";`;
				});

				// 添加 cool 导入语句
				importStatements.unshift(
					`import { coolPlugin as __coolPlugin } from "./.cool/bootstrap";`
				);

				// 找到 export function createApp() 的位置
				const appUvueIndex = getCodeIndex(/export\s+function\s+createApp\(\)/, code);

				if (appUvueIndex != -1) {
					// 在前面添加导入语句
					code =
						code.slice(0, appUvueIndex) +
						importStatements.join("\n") +
						"\n" +
						code.slice(appUvueIndex);

					// 生成插件使用语句
					const useStatements = pluginNames.map((pluginName) => {
						return `__${pluginName}Plugin.install(app);`;
					});

					// 添加 cool 使用语句
					useStatements.unshift(`__coolPlugin(app);`);

					// 找到 createSSRApp(App) 的位置，在其后插入插件使用代码
					const createSSRAppIndex = getCodeIndex(/createSSRApp\(App\);/s, code, true);

					if (createSSRAppIndex != -1) {
						// 在指定位置插入插件使用代码
						code =
							code.slice(0, createSSRAppIndex) +
							"\n" +
							useStatements.join("\n") +
							code.slice(createSSRAppIndex);
					}
				}

				return {
					code,
					map: { mappings: "" }
				};
			}
		}
	};
}

function uvuePlugin(): Plugin {
	return {
		name: "vite-cool-unix-uvue",
		enforce: "pre",
		transform(code, id) {
			if (id.endsWith("/App.uvue")) {
				// 找到 <script lang="uts"> 的位置
				const scriptIndex = getCodeIndex(/<script lang="uts">/, code) + 19;
				if (scriptIndex != -1) {
					// 在后面添加导入语句
					code =
						code.slice(0, scriptIndex) + '\nimport "./.cool";' + code.slice(scriptIndex);
				}

				code =
					code +
					`<style lang="scss">
      @import "@/.cool/index.scss";
       </style>`;

				return {
					code,
					map: { mappings: "" }
				};
			}
		}
	};
}

function localePlugin(): Plugin {
	return {
		name: "vite-cool-unix-locales",
		enforce: "pre",
		transform(code, id) {
			if (id.endsWith("/plugins/locale.ts")) {
				// 找到所有的 locales 文件
				const files = globSync(rootDir("/") + "**/locales/*.json", {
					withFileTypes: true
				});

				const importStatements: string[] = [
					`import { appendLocale as __appendLocale } from "@/.cool";`
				];
				const useStatements: string[] = [];

				// 遍历所有 locales 目录下的 JSON 文件，自动生成 import 和 appendLocale 语句
				for (const file of files) {
					// 确保是文件类型
					if (!file.isFile()) continue;

					// 获取父级目录名（假如没有则为 base），将中划线替换为下划线，避免变量非法
					const pName = file.parent?.parent?.name?.replace(/-/g, "_") || "base";

					// 文件相对路径，统一使用正斜杠
					const relativePath = file.path.replace(/\\/g, "/").replace(rootDir("/"), "");

					// 仅处理 .json 文件
					if (!file.name.endsWith(".json")) continue;

					// 文件名去后缀，变量名同样替换 - 为 _
					const fileName = basename(file.name, ".json");
					const varName = `loale_${pName}_${fileName.replace(/-/g, "_")}`;

					// 添加导入声明（import xxx from ...）
					importStatements.push(
						`import ${varName} from "@/${relativePath}/${file.name}";`
					);

					// 生成 appendLocale 语句
					const useStatement = `__appendLocale("${fileName}", ${varName});`;

					// 若是 uni_modules 路径，import 排到后面，非 uni_modules 路径优先
					if (relativePath.includes("uni_modules")) {
						useStatements.push(useStatement);
					} else {
						useStatements.unshift(useStatement);
					}
				}

				// 添加导入语句
				code = importStatements.join("\n") + "\n" + code;

				// 找到 install(app) { 的位置
				const installIndex = getCodeIndex(/install\s*\(\s*app\s*\)\s*\{/s, code, true);

				if (installIndex != -1) {
					// 在指定位置插入 appendLocale 使用语句
					code =
						code.slice(0, installIndex) +
						"\n" +
						useStatements.join("\n") +
						code.slice(installIndex);
				}

				return {
					code,
					map: { mappings: "" }
				};
			}
		}
	};
}

export function codePlugin(): Plugin[] {
	return [jsonPlugin(), tsPlugin(), uvuePlugin(), localePlugin()];
}
