// @cool-vue/unix/src/ctx/index.ts
import { join } from "path";
import { readFile, rootDir, writeFile } from "../utils";
import { glob } from "glob";
import { assign, cloneDeep, orderBy } from "lodash";
import util from "util";
import type { Plugin } from "vite";
import { getTailwindColor } from "../tailwind/utils";

export async function createCtx() {
	let ctx: Ctx.Data = {};

	const manifest = readFile(rootDir("manifest.json"), true);

	// 文件路径
	const ctxPath = rootDir("pages.json");

	// 页面配置
	ctx = readFile(ctxPath, true);

	// 原数据，做更新比较用
	const ctxData = cloneDeep(ctx);

	// 删除临时页面
	ctx.pages = ctx.pages?.filter((e) => !e.isTemp);
	ctx.subPackages = ctx.subPackages?.filter((e) => !e.isTemp);

	// 删除不需要的数据
	for (const i in ctx) {
		if (!["pages", "subPackages", "tabBar", "globalStyle", "uniIdRouter"].includes(i)) {
			delete ctx[i];
		}
	}

	// 加载 uni_modules 配置文件
	const files = await glob(rootDir("uni_modules") + "/**/pages_init.json", {
		stat: true,
		withFileTypes: true
	});

	for (const file of files) {
		if (file.isFile()) {
			const { pages = [], subPackages = [] }: Ctx.Data = readFile(
				join(file.path, file.name),
				true
			);

			// 合并到 pages 中
			[...pages, ...subPackages].forEach((e) => {
				e.isTemp = true;

				const isSub = !!e.root;

				const d = isSub
					? ctx.subPackages?.find((a) => a.root == e.root)
					: ctx.pages?.find((a) => a.path == e.path);

				if (d) {
					assign(d, e);
				} else {
					if (isSub) {
						ctx.subPackages?.unshift(e);
					} else {
						ctx.pages?.unshift(e);
					}
				}
			});
		}
	}

	// 排序后检测，避免加载顺序问题
	function order(d: Ctx.Data) {
		return {
			pages: orderBy(d.pages, "path"),
			subPackages: orderBy(d.subPackages, "root")
		};
	}

	// 是否需要更新 pages.json
	if (!util.isDeepStrictEqual(order(ctxData), order(ctx))) {
		console.log("[cool-ctx] pages updated");
		writeFile(ctxPath, JSON.stringify(ctx, null, 4));
	}

	// appid
	ctx.appid = manifest.appid;

	return ctx;
}

export function ctxPlugin(): Plugin {
	return {
		name: "vite-cool-unix-ctx",
		enforce: "pre",
		async transform(code, id) {
			if (id.includes("/ctx/index.ts") && id.includes("cool/")) {
				const ctx = await createCtx();

				// 主题配置
				const theme = readFile(rootDir("theme.json"), true);

				// 主题配置
				ctx["theme"] = theme || {};

				// 颜色值
				ctx["color"] = getTailwindColor();

				if (!ctx.subPackages) {
					ctx.subPackages = [];
				}

				if (!ctx.tabBar) {
					ctx.tabBar = {};
				}

				if (!ctx.uniIdRouter) {
					ctx.uniIdRouter = {};
				}

				let ctxCode = JSON.stringify(ctx, null, 4);

				ctxCode = ctxCode.replace(`"tabBar": {}`, `"tabBar": {} as TabBar`);
				ctxCode = ctxCode.replace(`"subPackages": []`, `"subPackages": [] as SubPackage[]`);

				code = code.replace("const ctx = {}", `const ctx = ${ctxCode}`);

				code = code.replace(
					"const ctx = parse<Ctx>({})!",
					`const ctx = parse<Ctx>(${ctxCode})!`
				);
			}

			if (id.endsWith(".json")) {
				const d = JSON.parse(code);

				// 转字符串，不然会报错：Method too large
				if (id.includes("/locales/")) {
					let t: string[] = [];

					(d as string[][]).forEach(([a, b]) => {
						t.push(`${a}<__=__>${b}`);
					});

					code = JSON.stringify([[t.join("<__&__>")]]);
				} else {
					code = JSON.stringify(d);
				}
			}

			return {
				code,
				map: { mappings: "" }
			};
		}
	};
}
