// @cool-vue/unix@1.1.0-beta.6/src/config.ts
export const config = {
  reqUrl: "",
  tailwind: {
    enable: true,
    remUnit: 16,
    remPrecision: 6,
    rpxRatio: 2,
    darkTextClass: "dark:text-surface-50",
  },
  isPlugin: false,
  utsPlatform: "web",
};

/**
 * 特殊字符映射表
 */
export const SAFE_CHAR_MAP: Record<string, string> = {
  "[": "-bracket-start-",
  "]": "-bracket-end-",
  "(": "-paren-start-",
  ")": "-paren-end-",
  "{": "-brace-start-",
  "}": "-brace-end-",
  $: "-dollar-",
  "#": "-hash-",
  "!": "-important-",
  "/": "-slash-",
  ":": "-colon-",
};

/**
 * 特殊字符映射表（国际化）
 */
export const SAFE_CHAR_MAP_LOCALE: Record<string, string> = {
  "[": "-bracket-start-",
  "]": "-bracket-end-",
  "(": "-paren-start-",
  ")": "-paren-end-",
  "{": "-brace-start-",
  "}": "-brace-end-",
  $: "-dollar-",
  "#": "-hash-",
  "!": "-important-",
  "/": "-slash-",
  ":": "-colon-",
  " ": "-space-",
  "<": "-lt-",
  ">": "-gt-",
  "&": "-amp-",
  "|": "-pipe-",
  "^": "-caret-",
  "~": "-tilde-",
  "`": "-backtick-",
  "'": "-single-quote-",
  ".": "-dot-",
  "?": "-question-",
  "*": "-star-",
  "+": "-plus-",
  "-": "-dash-",
  _: "-underscore-",
  "=": "-equal-",
  "%": "-percent-",
  "@": "-at-",
};
